Feature/marketplace suggestions

haszari opened the pull request:
### Please re-review this PR…
This is a clean new PR, in woocommerce repo, for the code in https://github.com/woocommerce/woocommerce/pull/22712 (custom fork, PR now closed).

The feedback from that PR's reviews has been addressed. Please refer to the above closed PR for any review comments / discussion. 

When this PR is reviewed and approved, we'll use this branch/PR as the feature branch for Marketplace Suggestions. The follow up PRs will be merged into here once they pass review.

Apologies for the confusion – any questions/problems ping me @haszari and I'll sort it out. Thanks for your help!

cc @kloon @mikejolley @justinshreve @kovshenin @timmyc

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Feature branch for Marketplace Suggestions. PRs relating to this feature will land here as they get approved.

Adds support for displaying subtle and appropriate in-product suggestion messaging for WooCommerce Marketplace extensions. These messages are driven by data from an API on WooCommerce.com.

### How to test the changes in this Pull Request:

* Checkout this branch and get it set up as the WooCommerce plugin in your development environment.
* Log in as a store administrator and view suggestions in the products admin screen – there are different suggestions displayed in the empty state and within the list of products.
* Test dismissing suggestions; dismissed suggestions should not show again.

This can also be tested on a [jurassic.ninja site](https://married-white-clawed.jurassic.ninja) - ping me if you want access.

haszari committed:
in-product discovery - render container for product-list suggestions

haszari committed:
in-product discovery - render suggestions in products list

haszari committed:
in-product discovery - don't promote extensions that are already installed:
- + move suggestions client code to a module, enqueued only on relevant pages

haszari committed:
register underscore dependency for marketplace script

haszari committed:
in-product discovery - show items if companion extension(s) installed

haszari committed:
in-product discovery - enqueue styles as separate module

haszari committed:
fix reference to data global (lint)

haszari committed:
don't use ES6 unnecessarily in gruntfile - breaks the linter/build

haszari committed:
remove use of ES6 features in marketplace suggestions (const, =>)

haszari committed:
style marketplace cards grid in empty products screen

haszari committed:
style marketplace suggestions

haszari committed:
fix issue with footer overlapping empty products layout - hide footer

haszari committed:
style marketplace suggestions list banner

haszari committed:
use attributes & DOM api for table row banner render - html snippet is too flexible/risky

haszari committed:
use attributes + DOM (not html snippet) for products empty state suggestions

haszari committed:
ensure list suggestions render correctly with optional url/copy attributes

haszari committed:
styling products empty state OBW style, including existing add/import + suggestions

haszari committed:
fill out products empty state suggestions:
- add back product suggestions data
- hide page admin elements, focus layout on onboarding
- bulk-fix indentation in admin.scss
- reinstate default BlankState styling (overrides are intended for products only, accidentally overwrote)

haszari committed:
products blank state - OBW style suggestions (missed in previous commit)

haszari committed:
move hard-coded suggestion data to woo API (WCCOM api coming soon)

haszari committed:
pull suggestion data from (proposed) S3 endpoint:
- leave hard-coded data in place until endpoint is available

haszari committed:
update styling of products table suggestion + add dismiss button

haszari committed:
ajax api for dismissing specific marketplace suggestions (first cut)

haszari committed:
ensure suggestion dismiss API / nonce is provided to js (missed in last commit)

haszari committed:
hide suggestions immediately after user clicks x (using jQuery.fadeOut)

haszari committed:
support dismissing empty-state suggestions + polish dismiss styling:\n- hover style for x\n- disable border/underline active state inherited from <a>\n- layout styling for empty-state suggestions, button on right\n- make test data more like mockup (no category recommendation in list)

haszari committed:
update products list empty state (incl. suggestions) to latest designs

haszari committed:
add header & footer to products empty suggestions box

haszari committed:
style products empty header/footer + support link style:
- added link-text to trigger link style for suggestion CTA, button is now opt-in
- made suggestion rendering more generic, should display something sensible if there are missing attribs
- all CTA links/buttons open in new tab via target attr

haszari committed:
add linkout icon to link-style suggestion CTAs (for empty products footer link)

haszari committed:
allow dismiss by default + hide products empty-state box if no suggestions

haszari committed:
mobile styles for products empty state suggestions

haszari committed:
fix button regression in product table suggestion

haszari committed:
share styles and markup for prpducts list suggestion

haszari committed:
hide inline products list suggestion on mobile (can't get the styling right)

haszari committed:
display icon images in suggestions + styling tweaks for header/footer

haszari committed:
fix size of products empty state icon in IE - use pixel sizing

haszari committed:
send tracks events when suggestions are displayed, clicked or dismissed:
- Using temporary stand-in tracks API until PR is merged.
- Also fixed logic so table-view suggestion is not "displayed" if
  empty-state suggestions are shown.

haszari committed:
pull real suggestions data from S3, remove hard-coded test data

haszari committed:
remove dismissed suggestions from DOM so styling updates accordingly

haszari committed:
cache empty suggestions if API fails, to reduce congestion/thrashing +
+ don't use wp_remote_safe_get – overkill
+ use wp_send_json_success to streamline/simplify API data ajax handler

haszari committed:
remove some extra whitespace

haszari committed:
fix client side data access after change to `wp_send_json_success`

haszari committed:
move woo-plugin slugs into a helper routine + cache installed woo plugins in a transient

haszari committed:
store dismissed suggestions in user meta, don't manually json-encode

haszari committed:
tidy up main marketplace suggestions module, make it a real class +
+ require it for admin only
+ fix issue with retrieving dismissed suggestions user meta value

haszari committed:
move suggestions class into admin/ and require from admin::includes

haszari committed:
move products list suggestions container render into module behind hook + move template near module, don't use wc_get_template to render

haszari committed:
support global opt-out of suggestions via user option 'woocommerce_allow_marketplace_suggestions'

haszari committed:
only display suggestions for Englishish locales

haszari committed:
document all js methods

haszari committed:
prefix AJAX actions to reduce chance of collision

haszari committed:
frontload suggestion data as script data, remove ajax request

haszari committed:
use updated tracks API wcSettings.recordEvent => wcTracks.recordEvent

haszari committed:
hide empty-state suggestions for mobile too – focus on desktop for v1

haszari committed:
only enqueue marketplace suggestions styles if they are enabled

haszari committed:
removed redundant `get_plugin_slug`

haszari committed:
Cache local woo plugins info rather than slugs-only subset:
- cache woo_plugins object for more general use
- streamline get_local_woo_plugin_slugs using pluck

haszari committed:
Improve clarity and logic for disabling suggestions:

- Use a filter rather than option
- Use return-early style in function for clarity
- Ensure user has 'install_plugins' capability

haszari committed:
MIA suggestions data should be cached as a real array, not text

haszari committed:
JS code-style fixes

haszari committed:
Merge branch master into feature/marketplace-suggestions

# Conflicts:
#assets/css/admin.scss

haszari committed:
(temporary) add hard-coded fake orders-list suggestion data

haszari committed:
(temporary) add hard-coded fake orders-list suggestion data

haszari committed:
allow suggestions on orders list + move logic into suggestions class

haszari committed:
allow suggestions on orders list + move logic into suggestions class

haszari committed:
render suggestions containers (& unstyled suggestions) in orders list

haszari committed:
style orders list suggestions + remove debug code

kovshenin committed:
Marketplace suggestions in Edit Product screen

Adds a new tab to Woo's product data metabox and renders the
appropriate containers for JS to inject some Marketplace
suggestions.

kovshenin committed:
Remove hard-coded JSON test data

kovshenin committed:
Add missing HTML template that renders the product data panel content

kovshenin committed:
Remove debug code

kovshenin committed:
Change marketplace suggestions label to "Get more options"

kovshenin committed:
Update Marketplace suggestions data using WC_Queue

Schedules an update task to run as soon as possible to fetch the
suggestions data, and cache it for a week. Failures are cached for
a day.

haszari committed:
use full path when including container template, rename folder 'views'

haszari committed:
remove icon from marketplace suggestions product edit tab

haszari committed:
rename product edit suggestions contexts for consistency

haszari committed:
opt-in to standard styles for product edit metabox suggestions

haszari committed:
tweak layout of product edit suggestions to match designs

haszari committed:
fix up products empty state styles now row-padding is balanced

haszari committed:
remove hard-coded test suggestion data

haszari committed:
hide footer, screen options when show suggestions in orders empty state

haszari committed:
allow suggestions to target multiple contexts:

If an array of context strings is specified in json, the suggestion will
be available for display in all specified contexts.

haszari committed:
allow suggestions in multiple contexts:

- fix regression when hiding footer
- rename `visibleSuggestions` for clarity

haszari committed:
limit maximum of 5 suggestions displayed in any given context

haszari committed:
Merge branch 'feature/marketplace-suggestions' into add/orders-list-suggestions

haszari committed:
fix regression introduced in merge - visibleSuggestions renamed

haszari committed:
Merge branch 'feature/marketplace-suggestions' into add/product-edit-suggestions

# Conflicts:
#	includes/admin/marketplace-suggestions/class-wc-marketplace-suggestions.php

+ fix linter errors / warnings

haszari committed:
hide product metabox suggestions tab (etc) when there are no suggestions

haszari committed:
use safe get for API request, linter fixes, doc comments

haszari committed:
automatically add utm params to suggestion urls, based on context

haszari commented:
The CI is failing on phpcs, on two existing files that I've touched.

How do we normally handle this – should these issues all be fixed as part of this PR?

The two files:

```
FILE: ...ome/travis/build/woocommerce/woocommerce/includes/admin/class-wc-admin.php
FOUND 29 ERRORS AFFECTING 24 LINES
....
FILE: ...is/build/woocommerce/woocommerce/includes/admin/helper/class-wc-helper.php
FOUND 98 ERRORS AFFECTING 84 LINES
```

mikejolley commented on includes/admin/helper/class-wc-helper.php:
```diff
@@ -1046,12 +1051,23 @@ public static function get_local_woo_plugins() {
 			$woo_plugins[ $filename ] = $data;
 		}
 
+		set_transient( 'wc_helper_installed_woo_plugins', $woo_plugins, DAY_IN_SECONDS );
```
Just a thought; store this transient as long as possible, and flush when a plugin is activated.

https://codex.wordpress.org/Plugin_API/Action_Reference/activated_plugin

This will mean cache is used more frequently.

haszari replied:
This code's gone now, using `active_plugins` instead, which is more conventional, and means we don't need to make changes in helper. 8dbda1b9c962a93be7d17ffa8ac55c2639f641e1

mikejolley commented on includes/admin/marketplace-suggestions/class-wc-marketplace-suggestions.php:
```diff
@@ -0,0 +1,161 @@
+		// Parse the data to check for any errors.
+		// If it's valid, store structure in transient.
+		if ( ! is_wp_error( $raw_suggestions ) ) {
+			$suggestions = json_decode( wp_remote_retrieve_body( $raw_suggestions ) );
```
Does json_decode need second argument to force assoc array vs object here? Not sure, just asking.

haszari replied:
Correct – [the default is `bool $assoc = FALSE`](https://secure.php.net/manual/en/function.json-decode.php). This is fixed in the `WC_Queue` updater:

https://github.com/woocommerce/woocommerce/pull/22857/files#diff-7cb8cd3240aef8097b64fc2183b79fb1R66

mikejolley reviewed (approved):
Looks good. I added one question, and one performance/cache suggestion.

R/e code standards, usually yes, but I'd leave them in this case to keep the branch/PRs clean.

kloon reviewed (approved):
LGTM, apart from the suggestions Mike mentioned.

Regarding the PHPCS fixes, you can do that in this PR in a separate commit, if it is too much and chances of introducing bugs then it is okay to skip it.

timmyc commented:
@kloon and @mikejolley thanks for the reviews here 🙇 - should this be merged into `master` ( the current target of the PR ) or should we merge this to a new feature branch off of current `master`?

kovshenin commented:
Can we get the other open PRs merged into this one for one final review?

kloon commented:
We need to get all the PR's that forms part of this one merged into it and then have it re-reviewed. Unfortunately one of the confusing things around having multiple PR's open around the same functionality. We need to try and stick to a single PR for future features, just easier to review and test.

kloon committed:
Merge pull request #22830 from woocommerce/add/marketplace-updater-wc-queue

Update Marketplace suggestions data using WC_Queue

kloon committed:
Merge pull request #22859 from woocommerce/add/marketplace-suggestions-campaign-tracking

add UTM params to marketplace suggestion URLs

kloon committed:
Merge pull request #22811 from woocommerce/add/orders-list-suggestions

Show marketplace suggestions in orders list empty state

timmyc commented:
Thanks for the clarification @kloon - so effectively this PR is the feature branch then correct?

kloon commented:
> Thanks for the clarification @kloon - so effectively this PR is the feature branch then correct?
That is correct yes. There's a bunch of PR's that needs merging into this one still before we can do a propper review.

kovshenin commented on includes/admin/helper/class-wc-helper.php:
```diff
@@ -1007,13 +1007,18 @@ private static function _get_subscriptions_from_product_id( $product_id, $single
 			require_once ABSPATH . 'wp-admin/includes/plugin.php';
 		}
 
+		$woo_plugins = get_transient( 'wc_helper_installed_woo_plugins' );
```
This will probably cause problems with the Helper.

We have no reliable way of determining when a new plugin is installed, or when one is deleted, and the Helper relies on this data always being accurate, so caching it, even for just a day, is going to cause confusion when users purchase a new extension, but the Helper doesn't see it for up to 24h.

The Helper will only call this method in the plugin (extension/subscription) management context, much like WordPress core. By context I mean the extension management screen in Woo core, the plugin list in wp-admin, as well as the plugin update checks in Cron. Caching this data for those routines has no real benefit in my opinion.

The problem we're trying to solve is that, what has previously been called in plugin management only, is now being called on many other non-plugin management screens in wp-admin, so caching sounds like a good solution. If we do cache it however, let's try and do this "internal" to the marketplace suggestions module to avoid any side effects on the Helper code.

haszari replied:
Now fixed – I've reverted these changes, now we're using `active_plugins` option. 8dbda1b9c962a93be7d17ffa8ac55c2639f641e1

kovshenin reviewed (changes requested):
See comment about caching in `get_local_woo_plugins`.

haszari committed:
show/hide suggestions based on which extensions are active:

- was previously based on folder names of installed plugins

haszari committed:
revert suggestion-related changes in helper:

- no longer required, now using active plugins for suggestions
- get_local_woo_plugins needs to not be cached
- left plugin `slug` in get_local_woo_plugins values, may be useful

haszari committed:
Merge branch feature/marketplace-suggestions into add/product-edit-suggestions

# Conflicts:
#assets/css/admin-rtl.css
#assets/css/admin.css
#assets/css/marketplace-suggestions-rtl.css
#assets/css/marketplace-suggestions.css
#assets/css/marketplace-suggestions.scss
#assets/js/admin/marketplace-suggestions.js
#includes/admin/class-wc-admin-assets.php
#includes/admin/marketplace-suggestions/class-wc-marketplace-suggestions.php

haszari committed:
Merge pull request #22815 from woocommerce/add/product-edit-suggestions

 Marketplace suggestions in Edit Product screen

haszari committed:
marketplace suggestions styling polish:

- opt-in to shared CTA/button styles for orders empty state
- tweak table banner icon margin for perfect centering
- ensure CTA div doesn't get squished and break layout \(product metabox\)
- ensure title+copy is vertically centered in table banner \(remove last margin\)

haszari committed:
randomly select suggestions to display, to improve visibility of each one

haszari committed:
switch in WCCOM new (versioned) suggestions host/URL

haszari committed:
update suggestions JS code to use `active` (WCCOM endpoint has updated schema)

kovshenin committed:
Update Marketplace suggestions endpoint URL

haszari committed:
align product metabox `Get more options` with other tab image labels

haszari committed:
remove outer border in metabox suggestions, lighten separators

haszari committed:
fix margins & ensmallen heading in product table suggestion

haszari committed:
align product metabox `Get more options` with other tab image labels (regressed)

- fix css selector
- also removed redundant margin-top rule

haszari committed:
improve margins & line-spacing in metabox copy

haszari committed:
Merge pull request #22871 from woocommerce/add/marketplace-suggestions-wccom-endpoint

Add/marketplace suggestions wccom endpoint

haszari commented:
This PR is ready to go from my / Team Alpha point of view.

- There are no follow up PRs to merge into this PR.
- This is blocked on @kovshenin review; however, the feedback has been addressed in the PR.

@timmyc @kloon What's the next step for merging this?

haszari committed:
remove unused function arg `slug`

timmyc commented:
@haszari excellent really happy to see the PR at this place! Looking at [circleci](https://travis-ci.org/woocommerce/woocommerce/jobs/499987111) it appears there are some phpcs fixes needed still but nothing major - but we will likely want to resolve those. 

I'll give this another 👀 from the JS side of things and a final 👍 - but thinking @kloon or @mikejolley will give it the final 👍 overall as that was the plan we discussed after getting all the related branches merged into this one.

timmyc commented on assets/js/admin/marketplace-suggestions.js:
```diff
@@ -0,0 +1,411 @@
+		window.wcTracks.recordEvent = window.wcTracks.recordEvent  || function() { };
+
+		// Tracks events sent in this file:
+		// - marketplace_suggestion_displayed
```
💟thanks for this documentation right here, being able to easily locate where tracks are recorded is such a lovely thing.

haszari replied:
Let's all promise to keep it updated when things change 🤝

timmyc commented on assets/js/admin/marketplace-suggestions.js:
```diff
@@ -0,0 +1,411 @@
+			if ( isButton ) {
+				linkoutButton.classList.add( 'button' );
+			}
+			else {
```
Minor but should this be on line 109? `} else {`

timmyc commented on assets/js/admin/marketplace-suggestions.js:
```diff
@@ -0,0 +1,411 @@
+
+		// Render DOM elements for suggestion content.
+		function renderSuggestionContent( title, copy ) {
+			var left = document.createElement( 'div' );
```
Super minor but in this method and `renderSuggestionCTA` the var names of `left` and `right` which I'm guessing align with how they are ultimately displayed on the page seem a bit odd in the methods. I suppose something like `container` or `wrapper` might make more sense - kind of like how you call it a `row` in `renderTableBanner` below. Again super minor just made me pause when reading through the code.

haszari replied:
The perils of lazy refactoring. Will look at fixing this as long as I'm confident it doesn't destabilise things :)

timmyc commented on assets/js/admin/marketplace-suggestions.js:
```diff
@@ -0,0 +1,411 @@
+			}
+
+			if ( ! buttonText ) {
+				buttonText = 'Go';
```
This needs to be translated.

haszari replied:
Good spotting, this is potentially a dumb default anyway. I'll wrap it and change to the more common `Learn More`.

haszari replied:
This was one of those handy code review tips that revealed something less-than-ideal. I've now improved some related things:

- The default string is now `Learn More` – it's used for almost all content, so is a much better default
- The default applies for both button and link CTAs (was previously only button)
- The default CTA style is button, as the link is used much less often

timmyc reviewed (commented):
Couple of pretty minor comments around the js logic, otherwise looking great to me. I really appreciate all the comments in the code throughout - made it easy to grok and follow along with.

haszari committed:
localise default CTA text + code-style improvements:

- else style, fix stale/presumptuous variable names
- localise default CTA text
- use `Learn More` for default, more in line with our latest content
- switch CTA default to button - link style is less common

haszari committed:
increase margin under empty-state suggestions so they aren't hard-aligned to bottom

haszari commented:
Just to be clear @timmyc – this PR does not include a minified build of the JS changes (I couldn't get that working!).

I'd been hoping/assuming that there's a build process as part of the beta build/release (etc). If not, I'm happy to work with the experts to get my js changes built.

mikejolley commented:
@haszari So I still have some problems with this PR before merging which I'll list below :)

1. The product edit screen section has no icon cmopared to the other sections

2. The behaviour when dismissing recommendations on the product list screen is super annoying. You dismiss, then when you come back you're immediately suggested something else. This feels way too intrusive. When you dismiss this should probably dismiss all suggestions for that area, at least for a period of time, otherwise everytime you refresh you're presenting something new which makes the dismissal appear broken.

3. The dismiss X itself needs a title or tooltip to label it as 'dismiss'.

4. Once you've dismissed everything on the product screen there is nothing that replaces the area, e.g. it could show 'No suggestions at present' or something with a link to the extension store. I see it dissapears after refresh only.

5. The suggestion on the product listing appears in a rather jarring way. It could probably slide out of something? I've accidently clicked it a few times because it pushes down the rest of the products suddenly without warning.

6. The scripts need to be minified. If you install `grunt` and run `grunt dev` there are a bunch of warnings that need to be fixed before this will work.

```
assets/js/admin/marketplace-suggestions.js
     33 |                    'slug': suggestionSlug,
                                                   ^ Extra comma. (it breaks older versions of IE)
     51 |            }
                      ^ Missing semicolon.
    101 |            linkoutButton.onclick = function( event ) {
                                                       ^ 'event' is defined but never used.
    105 |            }
                      ^ Missing semicolon.
    168 |                container.appendChild( renderDismissButton( slug ) )
                                                                             ^ Missing semicolon.
     29 |                ajaxurl,
                         ^ 'ajaxurl' is not defined.
    410 |})( jQuery, marketplace_suggestions, ajaxurl );
                                              ^ 'ajaxurl' is not defined.
    410 |})( jQuery, marketplace_suggestions, ajaxurl );
                     ^ 'marketplace_suggestions' is not defined.
      1 |/* global marketplace_suggestions ajaxurl */
         ^ 'marketplace_suggestions ajaxurl' is defined but never used.
   assets/js/frontend/add-to-cart-variation.js
    360 |                                    new_attr_select.find( 'option' ).each( function( index, el ) {
                                                                                                     ^ 'el' is defined but never used.
    360 |                                    new_attr_select.find( 'option' ).each( function( index, el ) {
                                                                                              ^ 'index' is defined but never used.
    360 |                                    new_attr_select.find( 'option' ).each( function( index, el ) {
                                                                                    ^ Functions declared within loops referencing an outer scoped variable may lead to confusing semantics. ($, attr_val, variation_active)
    386 |                    new_attr_select.find( 'option.attached.enabled' ).each( function( index, el ) {
                                                                                                      ^ 'el' is defined but never used.
    386 |                    new_attr_select.find( 'option.attached.enabled' ).each( function( index, el ) {
                                                                                               ^ 'index' is defined but never used.
```

Overall, the dismiss functionality feels broken. TBH, being able to dismiss each individual suggestion doesn't really seem beneficial, and causes real annoyances on the listing screens.

- Being able to dismiss on the blank slates and even the 'get more options' tab on the products screen doesn't really seem needed to me. These screens are only shown for a short period of time, and the product edit one is not visible unless you go to that tab. Those suggestions are super easy to ignore without feeling the need to dismiss them.
- On the flip side, being able to dismiss the more intrusive ones is essentual. That is, suggestions that appear inline in product lists, and any future suggestions that are visible on the page when it first loads. However, right now it's per suggestion. So when you have lots of suggestions it feels like the dismiss buttons are broken.

My suggestions:

1.  Only allow dismiss when a notice is visible on page load, on a screen which is used multiple times i.e. product list screen, not the blank slate or product edit page section.

2. Dismiss the area, not the suggestion. And/Or if dismissing the suggestion, have a cool down period before another suggestion is thrown back up in exactly the same place. e.g. a day.

cc @warrendholmes @avivapinchas @peterfabian

kyleaparker commented:
>Only allow dismiss when a notice is visible on page load, on a screen which is used multiple times i.e. product list screen, not the blank slate or product edit page section.

I'd be ok with removing the dismiss on the products and orders empty state since they are temporary. On the product edit section I think it makes sense to individually dismiss the suggestions so we can keep showing new things. 

>Dismiss the area, not the suggestion. And/Or if dismissing the suggestion, have a cool down period before another suggestion is thrown back up in exactly the same place. e.g. a day.

I agree that we should wait before showing more and change the title of the product edit tab section to something like "No recommended extensions right now":

<img width="641" alt="extension discovery i2 2019-03-04 11-37-07" src="https://user-images.githubusercontent.com/712873/53747724-e5414f80-3e71-11e9-8a40-31653d879be5.png">

haszari committed:
fade in product list table-row suggestions, so it's less jarring

haszari committed:
streamline product edit metabox when no suggestions are visible:

- hide title (animate slide up)
- trigger this after dismiss, so we're not left with a heading
and no content

haszari committed:
fix jshint warnings

haszari committed:
minify marketplace-suggestions js file

haszari committed:
add a tooltip (Dismiss this suggestion) to clarify dismiss button

haszari commented:
Thanks for picking up those details @mikejolley ! 

> The dismiss X itself needs a title or tooltip to label it as 'dismiss'.
I've added a tooltip to clarify what it does – `Dismiss this suggestion`.

> Once you've dismissed everything on the product screen there is nothing that replaces the area
Good point. What I've done here is to slide out the heading if there are no suggestions.

I went with this approach as you can get in that state by dismissing the current suggestions, but you'll get more next time you're in there.

> The suggestion on the product listing appears in a rather jarring way.
Sliding out is problematic in a table; I couldn't get it to work. To make it less jarring it's now fading in.

> The scripts need to be minified. 
I fixed the jshint errors, thanks for bringing it to my attention! I manually ran the `uglify` step (`node_modules/grunt/bin/grunt uglify`). 

There are existing jshint issues in `assets/js/frontend/add-to-cart-variation.js`, I don't think it's appropriate for me to tackle those as part of this PR.

kyleaparker commented:
>The product edit screen section has no icon cmopared to the other sections

This was intentional to make it clear the section is different - happy to revisit if it’s causing confusion

mikejolley commented:
@kyleaparker No thats fine, just was not sure if it was intentionally left out.

haszari committed:
snooze suggestions in product list (for 24 hours) after a suggestion is dismissed

haszari commented:
> have a cool down period before another suggestion is thrown back up in exactly the same place. e.g. a day.

Have now implemented this – this little "snooze" interlude may make this stronger overall, thanks for the suggestion :)

timmyc commented on assets/js/admin/marketplace-suggestions.js:
```diff
@@ -0,0 +1,433 @@
+			row.dataset.suggestionSlug = slug;
+
+			var cell = document.createElement( 'td' );
+			cell.setAttribute( 'colspan', 10 );
```
Testing this out with a backup from one of my "live" sites, and encountered an issue with this logic here. I have the Facebook extension installed which adds a column to the table, but other extensions do a similar thing, and users can also toggle columns on/off. So this is what the suggestion looks like to me:

![colspan](https://user-images.githubusercontent.com/22080/53844883-fb2b3f00-3f5b-11e9-8e49-776489077b58.png)

We could lean on jquery a bit here to determine the right number of columns that are actively displayed:

`$( 'table.wp-list-table.posts thead th:not(.hidden)' ).length`

Which would give us the correct number of th's that are not currently hidden. I _think_ that would work. Along those lines too, but perhaps a bit more edge-case, is when users toggle column visibility, things would still get wonky, even if adjusting the colspan accordingly. We'd have to listen for those events and respond accordingly. Not sure if that is a critical problem or not, but something to consider.

timmyc replied:
Screen grab of toggling columns: https://drive.google.com/file/d/1IkapaDFhEFBGnB61D5kjXfpNv2Os9MNI/view?usp=sharing

haszari replied:
Thanks for finding that – will look at implementing the jQuery based fix.

haszari replied:
It's now calculating the `colspan` when the banner is initially displayed.

I've also added an event handler to catch when screen options change and recalculate it. There are probably some other cases we could catch – keeping it simple for now.

https://github.com/woocommerce/woocommerce/pull/22857/commits/f944da3eacaf72e381fb3a856dbe2e3f17612021

timmyc reviewed (commented):
Tested this out on a new data set / site with some extensions installed and hit one little bugger. Did verify that the dismissal of suggestions in the product listing worked as expected. Logged out, and logged back in and was not shown any suggestions still.

haszari committed:
fix jshint issue (const => var) + minify marketplace-suggestions

haszari committed:
Ensure product list suggestion works with different screen options

- calculate product-table suggestion colspan using number of columns
- was previously hard-coded

haszari committed:
refresh products list suggestion colspan when screen options change

haszari committed:
snooze product-list suggestions for 2 days after dismiss (was 1)

nicomollet commented:
This is going too far.
I don't need suggestions, my shop managers don't need them.
Make this optional, thank you.
